src/main/kotlin/com/equisoft/awsmocks/utils/KtorLoggingUtils.kt

```diff
@@ -1,46 +1,124 @@
-                    ${call.receiveText()}
-                    ###
-                """.trimIndent()
-            )
+    if (application.log.isDebugEnabled) {
+        trace { application.log.debug(it.buildText()) }
+
+        receivePipeline.intercept(ApplicationReceivePipeline.Before) { request ->
+            cacheRequestBody(request)
+        }
+
+        intercept(ApplicationCallPipeline.Call) {
+            logRequestBody()
+        }
+    }
+}
+
+private suspend fun PipelineContext<Unit, ApplicationCall>.logRequestBody() {
+    val content = call.receiveText()
+    call.application.log.debug("""
+            ${call.request.httpMethod} ${call.request.path()}
+            ### CONTENT
+            $content
+            ###
+        """.trimIndent())
+}
+
+@OptIn(KtorExperimentalAPI::class, ExperimentalStdlibApi::class)
+@SuppressWarnings("ThrowsCount")
+private suspend fun PipelineContext<ApplicationReceiveRequest, ApplicationCall>.cacheRequestBody(
+    request: ApplicationReceiveRequest
+) {
+    val type: KType = request.typeInfo
+    val cachedResult: CachedResult<*>? = call.attributes.getOrNull(LastCachedResult)
+    when {
+        cachedResult == null -> call.attributes.put(LastCachedResult, RequestAlreadyConsumedResult)
+        cachedResult === RequestAlreadyConsumedResult -> throw RequestAlreadyConsumedException()
+        cachedResult is CachedResult.Failure -> throw cachedResult.cause
+        cachedResult is CachedResult.Success<*> && cachedResult.type == type -> {
+            proceedWith(ApplicationReceiveRequest(type, cachedResult))
+            return
+        }
+    }
+
+    try {
+        val byteArray = cacheContentAsByteArray(cachedResult, request)
+        val incomingContent = byteArray?.let { ByteReadChannel(it) } ?: cachedResult ?: request.value
+
+        proceedWith(ApplicationReceiveRequest(type, incomingContent, reusableValue = true))
+
+        if (cachedResult == null || cachedResult !is CachedResult.Success) {
+            call.attributes.put(LastCachedResult, CachedResult.Success(typeOf<ByteArray>(), byteArray ?: ByteArray(0)))
         }
+    } catch (@SuppressWarnings("TooGenericExceptionCaught") cause: Throwable) {
+        call.attributes.put(LastCachedResult, CachedResult.Failure(type, cause))
+        throw cause
+    }
+}
+
+@OptIn(ExperimentalStdlibApi::class)
+private suspend fun PipelineContext<ApplicationReceiveRequest, ApplicationCall>.cacheContentAsByteArray(
+    cachedResult: CachedResult<*>?,
+    request: ApplicationReceiveRequest
+): ByteArray? {
+    var byteArray: ByteArray? = (cachedResult as? CachedResult.Success<*>)?.value as? ByteArray
+
+    val requestValue: Any = request.value
+    if (byteArray == null && requestValue is ByteReadChannel) {
+        byteArray = requestValue.toByteArray()
+        call.attributes.put(LastCachedResult, CachedResult.Success(typeOf<ByteArray>(), byteArray))
     }
+
+    return byteArray
+}
+
+sealed class CachedResult<T : Any>(val type: KType) {
```

Comment:
je sais pu c'est quoi qui est "up to date", mettre des `{}` ou juste listé les classes une en dessous de l'autre?